Use `em_delete` in `key_delete`

etiennebarrie opened the pull request:
Since #434, the <kbd>Del</kbd> key is bound to `:key_delete`, but `:key_delete` doesn't behave like readline does in emacs mode. It moves the cursor left if a character was deleted under the cursor, while in readline, it just deletes the character under the cursor.

line | type | `key_delete` | `em_delete`
-|-|-|-
| `abc\|` | <kbd>DEL</kbd> | `abc\|` | same
| `ab\|c` | <kbd>DEL</kbd> | `a\|b` | `ab\|`
| `ab\|c` | <kbd>DEL</kbd> <kbd>DEL</kbd> | `\|a` | `ab\|`

`:em_delete` is almost <kbd>Del</kbd> like readline, except it also handles the end-of-file situation. Which is annoying because if you type <kbd>Del</kbd> on an empty line, you get EOF instead of no changes.

This fixes it by using `:em_delete` in emacs mode, but also tweaking the behavior of `:em_delete` to only handle EOF when the key received is Ctrl-D. I hardcoded this but we could revisit that, see #501.

sshock commented on lib/reline/line_editor.rb:
```diff
@@ -2664,7 +2666,7 @@ def finish
 
   private def em_delete(key)
-    if (not @is_multiline and @line.empty?) or (@is_multiline and @line.empty? and @buffer_of_lines.size == 1)
+    if @line.empty? and (not @is_multiline or @buffer_of_lines.size == 1) and key == 4
```
We could change 4 to `"\C-D".ord` for more clarity.

etiennebarrie replied:
Good idea, thanks! Also rebased the branch.

tompng commented on lib/reline/line_editor.rb:
```diff
@@ -2651,7 +2653,7 @@ def finish
 
   private def em_delete(key)
-    if (not @is_multiline and @line.empty?) or (@is_multiline and @line.empty? and @buffer_of_lines.size == 1)
+    if @line.empty? and (not @is_multiline or @buffer_of_lines.size == 1) and key == "\C-D"
```
`"\C-d".ord`(lowercase d) is better because lowercase is used in this file.
Could you add a test to check that `\C-d` with empty line will end the input?
I think we can use `assert_line(nil)`

etiennebarrie replied:
Thank you for pointing that out, I found a test in `test/reline/test_key_actor_vi.rb` that was testing the same case: `\C-d` with an empty line ends the input, so I added it to `test/reline/test_key_actor_emacs.rb` as well.

etiennebarrie commented:
I added some tests. As a separate first commit, tests that were testing the existing behavior that we want to preserve, namely <kbd>Ctrl-D</kbd> ends editing with an empty line, but that <kbd>Del</kbd> does not. Then there's the small refactor, then for the last commit changing the behavior, I added a test checking the state of the cursor to demonstrate the readline behavior for <kbd>Del</kbd>. That test fails when it runs on main.

tompng commented:
Thank you for your contribution :+1: :tada:
